Add origin field to constraints

Virtlink opened the pull request:
This PR adds an `origin` field to constraints, tracking the syntactic constraint to which substitutions were applied that resulted in the current constraint.

AZWN commented on statix.solver/src/main/java/mb/statix/constraints/CCompiledQuery.java:
```diff
@@ -108,11 +149,12 @@ public StateMachine<ITerm> stateMachine() {
-    @Override public CCompiledQuery unsafeApply(Immutable subst) {
+    @Override public CCompiledQuery unsafeApply(ISubstitution.Immutable subst, boolean trackOrigin) {
         return new CCompiledQuery(
                 filter.unsafeApply(subst),
```
If you want to trace & compare dataWf's etc, `trackOrigin` needs to be passed here, down to the underlying rule body, as well.

AZWN replied:
And similar to other occurrences in this file.

Virtlink replied:
I added code to track origins into apply() on the underlying rules.

AZWN commented on statix.solver/src/main/java/mb/statix/constraints/CResolveQuery.java:
```diff
@@ -69,46 +84,75 @@ public CResolveQuery withArguments(
+
+    @Override public CResolveQuery apply(ISubstitution.Immutable subst, boolean trackOrigin) {
         return new CResolveQuery(
                 filter.apply(subst),
```
Again, might want to track Origins here as well

AZWN commented on statix.solver/src/main/java/mb/statix/constraints/IResolveQuery.java:
```diff
@@ -1,5 +1,6 @@
 package mb.statix.constraints;
 
+import mb.scopegraph.resolution.StateMachine;
```
Is this import needed?

Virtlink replied:
Removed.

AZWN commented on statix.solver/src/main/java/mb/statix/spec/ApplyRelaxed.java:
```diff
@@ -50,9 +50,9 @@ class ApplyRelaxed extends ApplyMode<VoidException> {
         final IConstraint appliedBody;
         if(safety.equals(Safety.UNSAFE)) {
-            appliedBody = rule.body().unsafeApply(matchResult.substitution()).withCause(cause);
+            appliedBody = rule.body().unsafeApply(matchResult.substitution(), true).withCause(cause);
```
I am wondering if there are cases where either mode of application is used in a source-to-source transformation. In that case, we should actually lift the `trackOrigin` parameter, and set it to `true` in the solvers only. Perhaps that is the safest to do anyway, to keep this code usable for other transformations in the future.

Virtlink replied:
I've lifted the `trackOrigin` parameter, so that the solver can call the `apply()` methods of the ApplyModes with `trackOrigin` set to true.

AZWN reviewed (commented):
Thank you for submitting this PR. Good to see these changes landing! I have a few questions/remarks:

First, I'm puzzled by the addition of the `withArgument` methods. What was your motivation for those?

Second, is there a reason to assume that `withXXX` are called with referentially equal arguments often?

Virtlink commented:
About the `withArguments()`: in many places in Statix, a new copy is created of a constraint but with slightly different arguments. For example, instead of many instances of:

```java
new CEqual(newTerm1, newTerm2, c.cause().orElse(null), c.message().orElse(null));
```

now you use:

```java
c.withArguments(newTerm1, newTerm2);
```

and the remaining fields (metadata such as cause, message, now origin) are preserved. This way I don't also have to preserve origin everywhere only the main arguments of the constraint are updated.

I have found instances where sometimes a completely new constraint instance was constructed while all the fields were equal, which is why I added the if-checks. (I found this only with `withArguments()`, but the check is cheap so I added it to `withCause()` and `withMessage()` too. And I estimate that this will have a cascading effect: since some instances are no longer re-created, other instances detect that the arguments are unchanged and also not re-created.

AZWN reviewed (commented):
Hmmm, okay, that makes sense. However, I then would like to do this refactoring properly:
- Apply it throughout the codebase
- Don't support (or only partially) on `CCompiledQuery`
- Hide the constructors 
- Insert validity checks for implicitly transferred fields (such as the critical edges)

Perhaps it is best to split that out in a separate PR. Could you do that?

AZWN commented:
Also, we should check how often the exact referential equality actually occurs. E.g., on a substitution that does not change a term, I think a new instance is given.

Virtlink commented on statix.solver/src/main/java/mb/statix/constraints/CCompiledQuery.java:
```diff
@@ -56,12 +58,18 @@ private CCompiledQuery(
+    public CCompiledQuery withArguments(QueryFilter filter, QueryMin min, QueryProject project, ITerm scopeTerm, ITerm resultTerm) {
+        return withArguments(filter, min, project, scopeTerm, resultTerm, this.stateMachine);
+    }
+
```
Removed.
